### DISCUSSION TITLE
Checkout discount still applied on product/s when sale is deleted or the catalogue is empty

### DISCUSSION THREAD

**Discussion by fmravic96** [saleor/saleor#14617] [Q&A] [answered]
Hello everyone! 

(Using saleor v3.16)

We are currently having problems with updating checkout line price/total price when the sale is deleted or the catalogue is emptied out.

The issue seems to arise from the **create_or_update_discount_objects_from_sale_for_checkout** function in the saleor.discount.utils.py where if the sale is not existent it won't update the CheckoutLineDiscount object at all. It does work fine if we update the sale value or add/modify the catalogue.

This was working in previous versions but the change in 3.14 (I think?) somehow made this not possible anymore.

Any help is appreciated, thank you :)

**Comment by timuric** [comment]
Hi @fmravic96 could you please provide steps to reproduce the issue?

**Reply by fmravic96** [reply]
Hey @timuric,

From what I understand this has apparently been fixed in 3.17 where the promotion was replaced with the sale and the logic for deleting CheckoutLineDiscount objects was added but it does not exist in 3.16

### Steps to reproduce this problem

1. Create a 50% discount on a specific product
2. Use checkoutCreate mutation to add a specific product

Response
![image](https://github.com/saleor/saleor/assets/28535502/9afc04e1-3210-4a75-b42c-b6151f91a888)

3. Delete the sale/catalog
4. Fetch the checkout after sale deletion

Response
![image](https://github.com/saleor/saleor/assets/28535502/c8af3e2c-d7fa-413b-abfe-91d6563c2521)

I was expecting to see the updated checkout with no discounted product but it still has the 50% discount applied. Have I missed something in the process? It works when I apply the logic for deleting the CheckoutLineDiscount on the checkout lines from the 3.17 update.

**Reply by LeenaSomanS** [reply]
@korycins @simon-online, could you provide some insights or suggestions regarding the scenario described by @fmravic96? They mentioned a situation related to checkout details, specifically mentioning that discounts and gifts are not updating in the checkout details until the checkout session expires.

**Comment by DidoMarchet** [comment]
Hi, we encountered the same issue.

Do you have any updates?

Thanks and kind regards, 
Davide

**Comment by TurkaljGoran** [comment]
Same here unfortunately. Any feedback on this?

**Comment by IKarbowiak** [comment, marked as answer]
Hi,

[The issue](https://github.com/saleor/saleor/issues/14002) is known, but it shouldn't appear anymore on version 3.17. So if it's possible I recommend upgrading to version 3.17+. 🙂

**Comment by simon-online** [comment]
This still seems to be an issue in core v3.19.18.

When I delete a discount there is no worker task event logged so all the relevant discounted product pricing remains.

**Reply by korycins** [reply]
Hey @simon-online 

We've rebuilt the logic responsible for re-calculating the prices. Previously, the peak of changes applied to products, promotions, or sales could generate a huge amount of celery tasks that would process the same objects more than once. Here is the PR with these changes: https://github.com/saleor/saleor/pull/15051

Celere Beat schedules the task once per 30 sec, to re-calculate the prices of products marked as "dirty". So you need to have an active celery beat which will schedule the task asynchronously. 

If you still have an issue with the prices, please create the gh issue with all the steps required to re-produce the problem.

**Reply by simon-online** [reply]
I'm just doing the standard Docker build that creates the api and worker containers.

I was seeing the `beat: Starting...` and `beat: Waking up in 59.99 seconds.` messages logged by the beat process with debug on but none of the periodic / scheduled tasks were getting picked up and regularly run when I ran beat with the following command:
`celery -A saleor --app=saleor.celeryconf:app beat --loglevel=debug`

After a quite a few hours of trying different things I realised the issue was that the `settings.py` file didn't set celery to use the custom Saleor scheduler by including the following line:
`CELERY_BEAT_SCHEDULER = "saleor.schedulers.schedulers.PersistentScheduler"`

This line should really be included in the `settings.py` file so the periodic task scheduler works by default. I also tried the `saleor.schedulers.schedulers.DatabaseScheduler` scheduler but it had errors so I couldn't use it.

Another setting I made was included the following additional environment variables to ensure the periodic tasks didn't expire too soon:
`BEAT_UPDATE_SEARCH_FREQUENCY=60 seconds`
`BEAT_PRICE_RECALCULATION_SCHEDULE=60 seconds`
